### ISSUE TITLE
memory location error:  init of LayerNorm with kaiming_uniform_(), xavier_uniform_().

### ISSUE THREAD

**Issue by IgorAherne** [pytorch/pytorch#91379]
### 🐛 Describe the bug

When initializing the `named_parameters()`, the code raises memory error once it reaches `"norm1.weight"`. 
It happens if we try to use `xavier_uniform_()` or `kaiming_uniform_()`.

```
class TestModuleImpl : public torch::nn::Module {
public:
    TestModuleImpl(int numTransformers, int transformerSizes){
        for(int i=0; i<numTransformers; ++i){
            std::string name = std::string("encoderTransformer")+std::to_string(i);
            auto transf = register_module(name, torch::nn::TransformerEncoderLayer(transformerSizes, 2));
            _transformers.push_back(transf);
        }
        for(auto m : modules(false)){  kaimingInit_Layer(m);  }
    }

    void kaimingInit_Layer( const std::shared_ptr<Module>& m ){
        for(auto& p : m->named_parameters()){
            if (p.key().find("weight") != std::string::npos){
                torch::nn::init::kaiming_uniform_(p.value()); // <------------crashes here when "norm1.wieght", Error at mem location.
            }
        }
    }
    std::vector<torch::nn::TransformerEncoderLayer> _transformers;
};
TORCH_MODULE(TestModule);
```

### Versions

build version 1.13.0+cu117  
C++14
Debug and Release

cc @jbschlosser @bhosmer @cpuhrsch @erichan1

**Comment by IgorAherne** [comment]
Hey guys, any update on this, or what might be causing it?

**Comment by IgorAherne** [comment]
Actually, it appears the issue is not with transformer, but with the LayerNorm.

doing this in constructor causes exception to be raised:

torch::nn::LayerNormOptions lnOptions1({inputNumFeatures});
torch::nn::LayerNormOptions lnOptions2({inputNumFeatures});
_layerNorm1  = register_module("LayerNorm1", torch::nn::LayerNorm(lnOptions1));
_layerNorm2  = register_module("LayerNorm2", torch::nn::LayerNorm(lnOptions2));


void kaimingInit( const std::shared_ptr<Module>& m ){
        for(auto& p : m->named_parameters()){
            if (p.key().find("weight") != std::string::npos){
                torch::Tensor& val = p.value();
                torch::nn::init::kaiming_uniform_(p.value());  // <--------exception happens here, for LayerNorm
            }
            if (p.key().find("bias") != std::string::npos){  torch::nn::init::zeros_(p.value());  }
        }
    }

**Comment by IgorAherne** [comment]
Ok, I did an experiment and found a solid fix. 

Creating a custom implementation of LayerNorm **still was causing this error**, which was super weird.

This was causing the memory error:
```
//constructor
LayerNorm_harvImpl(int64_t features, double eps = 1e-6)
		: _weight(register_parameter("weight", torch::ones({features }))),
		_bias(register_parameter("bias", torch::zeros({features }))),
		_eps(eps) {}
```

Changing it from {features} to {1, features} suddenly resolved all the memory crashes.

```
//constructor
LayerNorm_CustomImpl(int64_t features, double eps = 1e-6)
		: _weight(register_parameter("weight", torch::ones({ 1, features }))),
		_bias(register_parameter("bias", torch::zeros({ 1, features }))),
		_eps(eps) {}
```

I suspect the implementation in the Libtorch library has the same issue so it might be very easy to fix for you.

**Comment by IgorAherne** [comment]
For further visitors: 

Make sure you are using **try{} catch(c10::Error& err)** blocks! It will tell you the error which is more descriptive than what seams like a memory crash :)

After I wrapped the code in try/catch I found out that `kaiming_init()` indeed didn't like the dimensions of the `LayerNorm.weight` the `err.what()` contained the message `"Fan in and fan out can not be computed for tensor with fewer than 2 dimensions"`

and wrapping the `torch::layer_norm()` in try/catch revealed `"Expected weight to be of same shape as normalized_shape, but got weight of shape [1, 64] and normalized_shape = [64]"`. 

This should get me back on track. I need to keep the weight shape as {features} rather than {1,features} and stop using kaiming init for the LayerNorm.
